challenger/src/grinding_challenger.rs

```diff
@@ -0,0 +1,34 @@
+use p3_field::PrimeField64;
+use p3_symmetric::CryptographicPermutation;
+
+use crate::{DuplexChallenger, FieldChallenger};
+
+pub trait GrindingChallenger<F: PrimeField64>: FieldChallenger<F> + Clone {
+    // Can be overridden for more efficient methods not involving cloning, depending on the
+    // internals of the challenger.
+    fn grind(&mut self, bits: usize) -> F {
+        for i in 0..F::ORDER_U64 {
+            let witness = F::from_canonical_u64(i);
+            let mut forked = self.clone();
+
+            if forked.check_witness(bits, witness) {
+                self.check_witness(bits, witness);
+                return witness;
+            }
+        }
+
+        panic!("failed to find witness")
+    }
+
+    fn check_witness(&mut self, bits: usize, witness: F) -> bool {
```

Comment:
Sorry for the all the nits, I'd just prefer it we keep `#[must_use]` to make sure the result can't be accidentally ignored. For the call above, we could do `let _ = ...`, or `assert!(...)`.